### PR TITLE
chore(benchmarks): memory allocation per function during page-load [s…

### PR DESCRIPTION
**Pull request by vigneshshanmugam**
+ part of elastic/apm-agent-rum-js#26 
+ This PR uses the chrome allocation profiler to measure the memory allocation of top functions from the agent that might help identify memory leaks
+ Generates a new build with keeping the function names of the apm agent to help with identify the functions with more memory allocation. 

Adds some keys to the benchmark documents that matches with the function names that uses the most memory 

```
   "memory.createResourceTimingSpan.bytes": 65664,
    "memory.generateRandomId.bytes": 33280,
    "memory.createResourceTimingSpans.bytes": 32864,
    "memory.getCurrentScript.bytes": 32816,
    "memory.patchFetch.bytes": 32816,
    "memory.baseExtend.bytes": 32784,
    "memory.isCORSSupported.bytes": 38376,
```

### CODE REVIEW

File: packages/rum/test/benchmarks/server.js [hunk 1]
```diff
@@ -93,4 +95,10 @@ module.exports = function startServer() {
       resolve(server)
     })
   })
-}
+})
+
+!(async () => {
+  if (require.main === module) {
```
Comment (vigneshshanmugam) [inline packages/rum/test/benchmarks/server.js]: added this so that we can just start the server for testing with `node packages/rum/test/benchmarks/server.js`

File: packages/rum/test/benchmarks/analyzer.js [hunk 2]
```diff
@@ -28,19 +28,48 @@ const stats = require('stats-lite')
...
+
+  return new Promise((resolve, reject) => {
+    webpack(config, err => {
+      if (err) {
+        reject(err)
+      }
+      console.info('custom apm build - ', filename, 'generated')
+      resolve()
+    })
+  })
```
Comment (jahtalab) [inline packages/rum/test/benchmarks/analyzer.js]: Why this needs to be async?

File: packages/rum/test/benchmarks/analyzer.js [hunk 3]
```diff
@@ -28,19 +28,48 @@ const stats = require('stats-lite')
...
+   */
+  const config = {
+    entry: path.join(__dirname, '../../src/index.js'),
+    output: {
+      filename,
+      path: path.join(dist, 'bundles'),
+      library: '[name]',
+      libraryTarget: 'umd'
+    },
+    ...getWebpackConfig(BUNDLE_TYPES.BROWSER_ESM_PROD)
```
Comment (jahtalab) [inline packages/rum/test/benchmarks/analyzer.js]: What do you think about moving the configs to dev-utils? e.g. having a separate function for getting the benchmark configs. One plus of this is that we can only set `keep_fnames: true ` on this configuration instead of changing the default production config.

File: packages/rum/test/benchmarks/analyzer.js [hunk 4]
```diff
@@ -181,12 +222,57 @@ function capturePayloadInfo(payload) {
...
+function getMemoryAllocationPerFunction({ profile }) {
+  const allocations = []
+  /**
+   * Whitelist built functions from the sample
+   */
+  const whitelist = [
+    '(V8 API)',
+    '(BYTECODE_COMPILER)',
+    '__webpack_require__',
+    'scriptId'
```
Comment (jahtalab) [inline packages/rum/test/benchmarks/analyzer.js]: What is the `scriptId`?

File: packages/rum/test/benchmarks/analyzer.js [hunk 5]
```diff
@@ -181,12 +222,57 @@ function capturePayloadInfo(payload) {
   }
 }
 
+function getMemoryAllocationPerFunction({ profile }) {
+  const allocations = []
+  /**
+   * Whitelist built functions from the sample
+   */
+  const whitelist = [
```
Comment (jahtalab) [inline packages/rum/test/benchmarks/analyzer.js]: I think you mean blacklist here, but even then I think something like `excludeFunctionNames` is better.

File: packages/rum/test/benchmarks/analyzer.js [hunk 6]
```diff
@@ -130,9 +160,20 @@ function calculateResults(resultMap) {
...
+       * Add consumed memory in bytes per each function to the result
        */
-      if (!Array.isArray(value)) {
+      if (metricName === 'memory') {
+        const reducedValue = value.reduce((acc, curr) => {
+          acc.push(...curr)
+          return acc
+        }, [])
+        reducedValue.forEach(obj => {
+          result[`memory.${obj.name}.bytes`] = obj.size
```
Comment (jahtalab) [inline packages/rum/test/benchmarks/analyzer.js]: This field will be parsed in Elasticsearch into hierarchy of objects, is that right?

File: packages/rum/test/benchmarks/profiler.js [hunk 7]
```diff
@@ -91,9 +103,17 @@ function gatherRawMetrics(browser, url) {
       Object.assign(metrics, timings)
     })
     /**
-     * Start the profiler before navigating to the URL
+     * Perform a garbage collection to do a clean check everytime
+     */
+    await client.send('HeapProfiler.collectGarbage')
```
Comment (jahtalab) [inline packages/rum/test/benchmarks/profiler.js]: Why is this necessary? Don't we start with a new window every time?

File: packages/rum/test/benchmarks/server.js [hunk 8]
```diff
@@ -39,28 +39,28 @@ function generateImageUrls(port, number) {
...
-function getRandomBundleContent() {
-  let content = APM_BUNDLE
+function getRandomBundleContent(apmBundle) {
+  let content = apmBundle
   content += `var scriptId = ${Date.now()};`
   return content
 }
 
-module.exports = function startServer() {
+const startServer = (module.exports = filename => {
```
Comment (jahtalab) [inline packages/rum/test/benchmarks/server.js]: I think it's cleaner to define the `startServer` function and then assign it to the exports.

File: packages/rum/test/benchmarks/analyzer.js (hunk 4 above)
Comment (vigneshshanmugam) [reply packages/rum/test/benchmarks/analyzer.js]: Its generated from webpack.

File: packages/rum/test/benchmarks/profiler.js (hunk 7 above)
Comment (vigneshshanmugam) [reply packages/rum/test/benchmarks/profiler.js]: Just to make sure memory is not reused, yes we do start a new window, but just in case.

File: packages/rum/test/benchmarks/analyzer.js (hunk 2 above)
Comment (vigneshshanmugam) [reply packages/rum/test/benchmarks/analyzer.js]: so we can await on the build part before starting the server.

File: packages/rum/test/benchmarks/analyzer.js (hunk 3 above)
Comment (vigneshshanmugam) [reply packages/rum/test/benchmarks/analyzer.js]: `build.js` is a common for all the builds. I prefer to keep the configs along with the files that are actually doing the job, we can definitely do the same `keep_fname: true` here also without overrding the default one. it would be the same wherever we keep this config.
